src/main/java/roboy/dialog/DialogStateMachine.java

```diff
@@ -0,0 +1,545 @@
+    // allows to have multiple instances of the same state class with different identifiers ("Greeting2" -> {GreetingState})
+    private HashMap<String, State> identifierToState;
+
+    private State activeState;
+    private State initialState;
+
+    // RosMainNode will be passed to every state as parameter
+    private final RosMainNode rosMainNode;
+    private final Neo4jMemoryInterface memory;
+
+    // Personality file additional information: everything like comment goes here.
+    // [!!] Do not use it in your State implementation! This info is only stored to make sure
+    //      we don't lose the comment etc. when saving this dialog state machine to file.
+    private HashMap<String, String> optionalPersFileInfo;
+
+    // endregion
+
+    //region constructors
+
+    /**
+     * Create an empty DialogStateMachine. Use loadFromFile(...) to load definitions from the
+     * personality file after creation. Alternatively you can create and add States to the machine
+     * manually from code.
+     *
+     * @param rmn reference to the RosMainNode that will be passed to every newly created State object
+     * @param mem reference to Memory that will be passed to every newly created State object
+     */
+    public DialogStateMachine(RosMainNode rmn, Neo4jMemoryInterface mem) {
+        identifierToState = new HashMap<>();
+        activeState = null;
+        rosMainNode = rmn;
+        memory = mem;
+        optionalPersFileInfo = new HashMap<>();
+
+        if (rosMainNode == null) {
+            logger.info("RosMainNode will be unavailable in DialogStateMachine (null was passed)");
+        }
+        if (memory == null) {
+            logger.info("Memory will be unavailable in DialogStateMachine (null was passed)");
+        }
+    }
+
+    /**
+     * Create an empty OFFLINE DialogStateMachine without a reference to the RosMainNode and Memory.
+     * States will not be able to access the RosMainNode and Memory functionality.
+     * This constructor is mainly used for testing.
+     */
+    public DialogStateMachine() {
+        this(null, null);
+    }
+
+    //endregion
+
+
+    // #####################################################
+    // #                 Getter & Setter                   #
+    // #####################################################
+
+    //region initial state
+
+    /**
+     * Returns the initial state for this state machine.
+     * @return initial state for this state machine
+     */
+    public State getInitialState() {
+        return initialState;
+    }
+
+    /**
+     * Set the initial state of this state machine.
+     * The state will be automatically added to the machine if not already added.
+     * If active state was null, it will be set to the new initial state.
+     * @param initial initial state
+     */
+    public void setInitialState(State initial) {
+        if (!identifierToState.containsValue(initial) && initial != null) {
```

Comment:
Maybe the other way around as there is no need to call the method if s == null

Reply:
good point!